topology/probes/istio/verifiers.go

```diff
@@ -0,0 +1,141 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package istio
+
+import (
+	kiali "github.com/kiali/kiali/kubernetes"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology/probes/k8s"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+)
+
+type verifier map[string][]func(*graph.Graph)
+
+func getDefaultVerifier() verifier {
+	var v verifier
+	v = make(map[string][]func(*graph.Graph))
+	return v
+}
+
+func (v verifier) setOnAdd(onAddFuncs ...func(*graph.Graph)) {
+	v["onAdd"] = onAddFuncs
+}
+
+func (v verifier) setOnUpdate(onUpdateFuncs ...func(*graph.Graph)) {
+	v["onUpdate"] = onUpdateFuncs
+}
+
+func (v verifier) setOnDelete(onDeleteFuncs ...func(*graph.Graph)) {
+	v["onDelete"] = onDeleteFuncs
+}
+
+// ResourceVerifier describes a set of verifiers for a single istio resource
+type ResourceVerifier struct {
+	*graph.EventHandler
+	*k8s.KubeCache
+	g *graph.Graph
+	v verifier
+}
+
+// OnAdd is called when a new Kubernetes resource has been created
+func (rv *ResourceVerifier) OnAdd(obj interface{}) {
+	rv.g.Lock()
+	defer rv.g.Unlock()
+	for _, v := range rv.v["onAdd"] {
+		v(rv.g)
+	}
+}
+
+// OnUpdate is called when a Kubernetes resource has been updated
+func (rv *ResourceVerifier) OnUpdate(oldObj, newObj interface{}) {
+	rv.g.Lock()
+	defer rv.g.Unlock()
+	for _, v := range rv.v["onUpdate"] {
+		v(rv.g)
+	}
+}
+
+// OnDelete is called when a Kubernetes resource has been deleted
+func (rv *ResourceVerifier) OnDelete(obj interface{}) {
+	rv.g.Lock()
+	defer rv.g.Unlock()
+	for _, v := range rv.v["onDelete"] {
+		v(rv.g)
+	}
+}
+
+// NewResourceVerifier returns a new cache using the associed Kubernetes
+// client and with the handler for the resource that this cache manages.
+func NewResourceVerifier(restClient rest.Interface, objType runtime.Object, resources string, g *graph.Graph, v verifier) *ResourceVerifier {
+	rv := &ResourceVerifier{
+		EventHandler: graph.NewEventHandler(100),
+		g:            g,
+		v:            v,
+	}
+	rv.KubeCache = k8s.RegisterKubeCache(restClient, objType, resources, rv)
+	return rv
+}
+
+func initResourceVerifiers(client interface{}, handlers []verifierHandler, g *graph.Graph) []probe.Probe {
+	verifiers := []probe.Probe{}
+	for _, h := range handlers {
+		verifiers = append(verifiers, h(client, g))
+	}
+	return verifiers
+}
+
+func findMissingGateways(g *graph.Graph) {
```

Comment:
this is part of either the `virtualservice` or `gateway` verification code and thus should be moved to say `vistualservice.go`. Such that all the code related to virtualservice (node, edge, verificaiton) is contained in that file.

Reply:
FIXED.